packages/@aws-cdk/pipelines/lib/codepipeline/codepipeline-engine.ts

```diff
@@ -0,0 +1,888 @@
+import * as path from 'path';
+import * as cb from '@aws-cdk/aws-codebuild';
+import * as cp from '@aws-cdk/aws-codepipeline';
+import * as cpa from '@aws-cdk/aws-codepipeline-actions';
+import * as ec2 from '@aws-cdk/aws-ec2';
+import * as iam from '@aws-cdk/aws-iam';
+import { Aws, Fn, IDependable, Lazy, PhysicalName, Stack } from '@aws-cdk/core';
+import * as cxapi from '@aws-cdk/cx-api';
+import { Construct, Node } from 'constructs';
+import { AssetType, FileSet, ManualApprovalStep, ScriptStep, StackAsset, StackDeployment, Step } from '../blueprint';
+import { DockerCredential, dockerCredentialsInstallCommands, DockerCredentialUsage } from '../docker-credentials';
+import { GraphNode, GraphNodeCollection, isGraph, AGraphNode, PipelineGraph } from '../helpers-internal';
+import { BuildDeploymentOptions, IDeploymentEngine } from '../main/engine';
+import { appOf, assemblyBuilderOf, embeddedAsmPath, obtainScope } from '../private/construct-internals';
+import { toPosixPath } from '../private/fs';
+import { enumerate, flatten, maybeSuffix } from '../private/javascript';
+import { writeTemplateConfiguration } from '../private/template-configuration';
+import { CodeBuildFactory, mergeCodeBuildOptions, stackVariableNamespace } from './_codebuild-factory';
+import { ArtifactMap } from './artifact-map';
+import { CodeBuildStep } from './codebuild-step';
+import { CodePipelineActionFactoryResult, ICodePipelineActionFactory } from './codepipeline-action-factory';
+
+/**
+ * Creation properties for a `CodePipelineEngine`
+ */
+export interface CodePipelineEngineProps {
+  /**
+   * The name of the CodePipeline pipeline
+   *
+   * @default - Automatically generated
+   */
+  readonly pipelineName?: string;
+
+  /**
+   * Create KMS keys for the artifact buckets, allowing cross-account deployments
+   *
+   * The artifact buckets have to be encrypted to support deploying CDK apps to
+   * another account, so if you want to do that or want to have your artifact
+   * buckets encrypted, be sure to set this value to `true`.
+   *
+   * Be aware there is a cost associated with maintaining the KMS keys.
+   *
+   * @default false
```

Comment:
This feels like something that can trip people over - can we make this a smart default and automatically enable if users try to deploy stacks across accounts?

Reply:
Doesn't feel safe, because that also means we'd automatically *remove* the key once they stop deploying across accounts (and that might lead to data loss as the key gets deleted).

The error message they get if they forget to turn it on is pretty clear.

Reply:
I think having to set this to `true` for any multi-environment application is unnecessary friction.

Can't we create the key with RETAIN removal policy?

Reply:
It doesn't feel too onerous, and opt-in seems like the right call to make here. Generally our customers don't appreciate it if we decide for them that cost isn't important.